instance/tests/models/test_server.py

```diff
@@ -125,51 +125,117 @@ def authenticate(client):
-    def test_sleep_until_status(self, mock_sleep, mock_update_status):
+    def test_sleep_until_condition_already_fulfilled(self, mock_sleep, mock_update_status):
         """
-        Sleep until the server gets to 'booted' status (single status string argument)
+        Check if sleep_until behaves correctly if condition to wait for
+        is already fulfilled.
         """
-        server = OpenStackServerFactory()
-        status_queue = [
-            None,
-            server._status_to_started,
-            server._status_to_active,
-            server._status_to_booted,
-            server._status_to_terminated,
+        conditions = [
+            lambda: server.status.is_steady_state,
+            lambda: server.status.accepts_ssh_commands,
+        ]
+        for condition in conditions:
+            server = OpenStackServerFactory()
+            status_queue = [
+                server._status_to_started,
+                server._status_to_active,
+                server._status_to_booted,
+            ]
+            # Transition to state fulfilling condition
+            for state_transition in status_queue:
+                server._transition(state_transition, progress=ServerProgress.Success)
+
+            # Sleep until condition is fulfilled
+            server.sleep_until(condition)
+            self.assertEqual(server.status, ServerStatus.Booted)
+            self.assertEqual(server.progress, ServerProgress.Success)
+            self.assertEqual(mock_sleep.call_count, 0)
+
+    @patch('instance.models.server.OpenStackServer.update_status')
+    @patch('instance.models.server.time.sleep')
+    def test_sleep_until_state_changes(self, mock_sleep, mock_update_status):
+        """
+        Check if sleep_until behaves correctly if condition to wait for
+        is unfulfilled initially.
+        """
+        conditions = [
+            lambda: server.status.is_steady_state,
+            lambda: server.status.accepts_ssh_commands,
         ]
-        status_queue.reverse() # To be able to use pop()
+
+        def scoped_update_status(server=None, status_queue=None):
+            """ Return mock update_status scoped to a specific server and status_queue """
+            def update_status():
+                """ Simulate status progression """
+                server._transition(status_queue.pop(), progress=ServerProgress.Success)
+            return update_status
+
+        for condition in conditions:
+            server = OpenStackServerFactory()
+            status_queue = [
+                server._status_to_started,
+                server._status_to_active,
+                server._status_to_booted,
+            ]
+            status_queue.reverse() # To be able to use pop()
+
+            mock_update_status.side_effect = scoped_update_status(server, status_queue)
+            mock_sleep.call_count = 0
+
+            # Sleep until condition is fulfilled
+            server.sleep_until(condition)
```

Comment:
@e-kolpakov Yes, you're right, it would wait until the default timeout has passed. I changed it to use a much smaller timeout (here and in a couple more places).